### ISSUE TITLE
Host Overview cannot marshall pmproxy response for some hosts

### ISSUE THREAD

**Issue by whitingjr** [performancecopilot/grafana-pcp#82]
**Data source:** PCP Redis

**Describe the bug**
Looking at the Host Overview Dashboard does not display metrics for some hosts. Out of 8 only 3 work as expected.
Error message in each panel is

`cannot unmarshal response: json: cannot unmarshal object into Go value of type []pmseries.ValuesResponseItem`

**To Reproduce**

1. Go to PCP Redis: Host Overview
2. In my case select 4th host in the drop down list box control
3. See error

**Expected behaviour**
Metrics on each panel are displayed.

**Screenshots**

Expected
![expected-screenshot-02](https://user-images.githubusercontent.com/708428/116696520-ed499b00-a9b9-11eb-9f58-832bc9889ea4.png)

Actual
![actual-screenshot-02](https://user-images.githubusercontent.com/708428/116696528-f0dd2200-a9b9-11eb-970c-c5edffba7d79.png)

**Versions (please complete the following information):**
 - Performance Co-Pilot:5.1.1
 - Grafana:7.5.4
 - grafana-pcp:3.0.3
- Redis:5.0.3

**Additional context**

```
# curl http://localhost:44322/pmapi/context
{"context":76585460,"source":"6dfd03eae783f24af2715adb244788fafa64de99","hostspec":"acmeserver","labels":{"domainname":"localdomain","hostname":"acmeserver","machineid":"5e130fd5d4c1492494ee75bd713b323e"}}
```

 How do I troubleshoot this ?

**Comment by andreasgerstmayr** [comment]
@whitingjr : can you paste the contents of `/var/log/grafana/grafana.log`? The error means pmproxy is sending JSON which grafana-pcp cannot parse, and the `grafana.log` contains the exact request to pmproxy. After getting the URL of this response, please run `curl` with it and paste the contents here.

Another logfile which may provide hints is `/var/log/pcp/pmproxy/pmproxy.log`.

**Comment by whitingjr** [comment]
pmproxy has little to go on
```
[Fri Apr 30 10:25:07] pmproxy(647745) Info: OpenSSL 1.1.1g FIPS  21 Apr 2020 - no certificates found
[Fri Apr 30 10:25:07] pmproxy(647745) Info: Redis slots, command keys, schema version setup
Fri Apr 30 10:55:01 2021 discovery callback: log-rolling in progress
Fri Apr 30 10:55:01 2021 discovery callback: finished log-rolling
...
Tue May  4 11:25:26 2021 discovery callback: log-rolling in progress
Tue May  4 11:25:27 2021 discovery callback: finished log-rolling
```
grafana has useful errors
```
t=2021-05-04T15:36:52+0000 lvl=eror msg="Cannot unmarshal response" logger=plugins.backend pluginId=pcp-redis-datasource code=200.000 err="json: cannot unmarshal object into Go value of type []pmseries.ValuesResponseItem" url="http://abc01.somewhere.com:44322/series/values?finish=1620142672&interval=30&series=1aacd03ae9ada8239c1619bd75eee17fd2986538&start=1620120892"
t=2021-05-04T15:36:52+0000 lvl=info msg="Request Completed" logger=context userId=5 orgId=1 uname=elmerfud@acme.com method=POST path=/api/ds/query status=400 remote_addr=10.36.115.83 time_ms=1 size=316 referer="https://somewhere.com:3443/d/8qt7la9Mz/pcp-redis-host-overview?orgId=1&refresh=10s&var-datasource=PCP%20Redis&var-host=abc04.somewhere.com"
t=2021-05-04T15:36:53+0000 lvl=dbug msg="Processing metrics query" logger=http.server query="&{data:map[datasource:PCP Redis datasourceId:4 expr:mem.vmstat.pgmajfault{hostname == \"abc04.somewhere.com\"} format:time_series intervalMs:30000 legendFormat:$metric maxDataPoints:572 refId:A]}"
t=2021-05-04T15:36:53+0000 lvl=dbug msg=handleTimeSeriesQueries logger=plugins.backend pluginId=pcp-redis-datasource request="map[Headers:<nil> PluginContext:map[AppInstanceSettings:<nil> DataSourceInstanceSettings:map[BasicAuthEnabled:false BasicAuthUser: Database: DecryptedSecureJSONData:<nil> ID:4 JSONData:map[] Name:PCP Redis UID: URL:http://abc01.somewhere.com:44322 Updated:2021-04-30T09:29:38Z User:] OrgID:1 PluginID:pcp-redis-datasource User:map[Email:elmerfud@acme.com Login:elmerfud@acme.com Name:Jeremy Whiting Role:Viewer]] Queries:[map[Interval:3e+10 JSON:map[datasource:PCP Redis datasourceId:4 expr:mem.vmstat.pgmajfault{hostname == \"abc04.somewhere.com\"} format:time_series intervalMs:30000 legendFormat:$metric maxDataPoints:572 refId:A] MaxDataPoints:572 QueryType: RefID:A TimeRange:map[From:2021-05-04T09:36:52.453Z To:2021-05-04T15:36:52.453Z]]]]"
t=2021-05-04T15:36:53+0000 lvl=dbug msg=Query logger=plugins.backend pluginId=pcp-redis-datasource query="map[expr:mem.vmstat.pgmajfault{hostname == \"abc04.somewhere.com\"} format:time_series legendFormat:$metric]"
t=2021-05-04T15:36:53+0000 lvl=dbug msg="HTTP Request" logger=plugins.backend pluginId=pcp-redis-datasource url="http://abc01.somewhere.com:44322/series/query?expr=mem.vmstat.pgmajfault%7Bhostname+%3D%3D+%22abc04.somewhere.com%22%7D"
t=2021-05-04T15:36:53+0000 lvl=dbug msg="Processing metrics query" logger=http.server query="&{data:map[datasource:PCP Redis datasourceId:4 expr:network.interface.in.bytes{hostname == \"abc04.somewhere.com\"} format:time_series intervalMs:30000 legendFormat:$instance maxDataPoints:572 refId:A]}"
```

```
# curl -v http://abc01.somewhere.com:44322/series/values?finish=1620142672&interval=30&series=1aacd03ae9ada8239c1619bd75eee17fd2986538&start=1620120892
[1] 1428604
[2] 1428605
[3] 1428606
[root@abc01 ~]# *   Trying xxx.xxx.xxx.xxx...
* TCP_NODELAY set
* Connected to abc01.somewhere.com (xxx.xxx.xxx.xxx) port 44322 (#0)
> GET /series/values?finish=1620142672 HTTP/1.1
> Host: abc01.somewhere.com:44322
> User-Agent: curl/7.61.1
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Connection: Keep-Alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Accept, Accept-Language, Content-Language, Content-Type
< Content-Length: 19
< Content-Type: application/json
< Date: Tue, 04 May 2021 15:41:40 GMT
< 
{"success":false}
* Connection #0 to host abc01.somewhere.com left intact

[1]   Done                    curl -v http://abc01.somewhere.com:44322/series/values?finish=1620142672
[2]-  Done                    interval=30
[3]+  Done                    series=1aacd03ae9ada8239c1619bd75eee17fd2986538
[root@abc01 ~]#
```
this is an extract with representative error logging.

**Comment by andreasgerstmayr** [comment]
Can you please re-run the curl command with the URL quoted (`curl -v "http://abc01.somewhere.com:44322/series/values?finish=1620142672&interval=30&series=1aacd03ae9ada8239c1619bd75eee17fd2986538&start=1620120892"`)? Without quoting, bash is splitting up the query at the `&` symbol (and running it in the background).

**Comment by andreasgerstmayr** [comment]
> At a guess the client is expecting a json with series metric value information and instead receiving a boolean outcome.

Spot on. This bug was fixed in PCP 5.2.0 (https://github.com/performancecopilot/pcp/commit/83d3466bcf9e796a00898b42c7f95a3a1b6eddd9).

Are you on RHEL/CentOS 8.3, but with the latest Grafana & grafana-pcp from upstream? iirc grafana-pcp before version 3 had a check for this faulty response, but due to the rewrite of the PCP Redis datasource in Golang in grafana-pcp v3 (adding alerting capabilities and other things), and because it's already fixed in PCP since some time, I must have missed to add this workaround in grafana-pcp 3.x.

But that's only a display error, the underlying issue is that some hosts apparently don't have any data in Redis. Can you please run one more command to verify this: `pmseries 'kernel.all.sysfork{hostname=="<your_host_name>"}[count:10]'`. It should list the last 10 metric values of the `kernel.all.sysfork` metric for the hostname `<your_host_name>`.

After confirming that Redis doesn't have metric values of this host stored, you can run `pmdumplog /var/log/pcp/pmlogger/<your_host_name>/<archive> kernel.all.sysfork ` to see if pmlogger actually captured data from these remote host or not. You can also run `pcp` to see which pmloggers are running.

**Comment by whitingjr** [comment]
@andreasgerstmayr All the component versions are in the initial [Versions](https://github.com/performancecopilot/grafana-pcp/issues/82#issue-872449216) section.
Running the command on the pmcd collector machine I see these entries
```
# pmseries 'kernel.all.load{hostname=="abc04.somewhere.com"}[count:10]'

dec61ed572e4312070dc7bbe9dc507980bce5dec
    [Wed May  5 11:43:20.952355000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:43:20.952355000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:43:20.952355000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:42:20.952302000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:42:20.952302000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:42:20.952302000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:41:20.952456000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:41:20.952456000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:41:20.952456000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:40:20.950394000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:40:20.950394000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:40:20.950394000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:39:20.950463000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:39:20.950463000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:39:20.950463000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:38:20.952465000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:38:20.952465000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:38:20.952465000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:37:20.952375000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:37:20.952375000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:37:20.952375000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
    [Wed May  5 11:36:20.950484000 2021] 0.000000e+00 2293aeb192cb8f37b5c2ab261e963e4d4f247b94
    [Wed May  5 11:36:20.950484000 2021] 0.000000e+00 41df8d3aa99ab59b20b8c1db03be7271a5b7567f
    [Wed May  5 11:36:20.950484000 2021] 0.000000e+00 93e56fbf627af525916636f8b7022818abfd205d
#
```
 For some reason it stopped getting logs yesterday from the machine. Not been able to troubleshoot this yet. The pmcd and pmlogger on the machine have a status of Thawing :/

**Comment by whitingjr** [comment]
> This bug was fixed in PCP 5.2.0
ok so until the newer version becomes available in the repo I am scuppered. Unless I can get it working in a container. Thank you for mentioning.

**Comment by whitingjr** [comment]
I will update to the latest PCP I can get my hands on and do further testing. Closing this issue for now.
